tests/ui/async-await/issue-76547.stderr

```diff
@@ -1,27 +1,25 @@
 error: lifetime may not live long enough
-  --> $DIR/issue-76547.rs:20:13
+  --> $DIR/issue-76547.rs:19:14
    |
 LL | async fn fut(bufs: &mut [&mut [u8]]) {
-   |                    -     - let's call the lifetime of this reference `'2`
-   |                    |
-   |                    let's call the lifetime of this reference `'1`
-LL |     ListFut(bufs).await
-   |             ^^^^ this usage requires that `'1` must outlive `'2`
+   |              ^^^^  -     - let's call the lifetime of this reference `'2`
+   |              |     |
+   |              |     let's call the lifetime of this reference `'1`
+   |              assignment requires that `'1` must outlive `'2`
```

Comment:
this seems wrong to me :thinking:

the assignment is very clearly not part of the function signature xx that's likely caused by the async fn desugaring first moving `bufs` into a local, and we then constrain the type of the local to be `&'a mut &'a` :/

I think we should somehow filter this, I feel like the current approach of figuring out what to blame is generally a bit :shrug: i would expect us to instead iterate over all constraints and take the "best one" (while preferring earlier ones), instead of simply taking the first interesting one

Reply:
ah, yeah, that's definitely wrong. I had a note about that test but I completely forgot to address it; thank you for the catch! I've added a commit that makes it try not to blame constraints from desugarings. it includes a special case to allow blaming returns from `?`, since otherwise [`tests/ui/async-await/issue-67765-async-diagnostic.stderr`](https://github.com/rust-lang/rust/blob/master/tests/ui/async-await/issue-67765-async-diagnostic.stderr) would be less clear about why the usage `Err(b)` is problematic (and `?` including a return is a lot more clear than the async sugar including an assignment). I'm not a desugaring expert, though, so I'm not sure if any others are intuitive/meaningful enough to be blamed (possibly with special wording?).

with regard to iterating over all constraints and picking the "best one", I agree that it should look at all the constraints in some way, but I'll have to think about if there's a better heuristic for the single best than position on the path (with some filter). I think returns are generally good to point to, and `CopyBound` constraints should maybe at least get an honorable mention in a note (if not the full blame) since they're subtle, but beyond that I think it gets a bit fuzzy. e.g. ascriptions and casts are helpful to point to when they syntactically refer to a relevant type/region, but are much less helpful if they leave it up to inference. for another example, `UseAsConst` is sometimes the best we've got at the top level, but within function bodies it's usually less helpful. besides the category, it feels like it should be possible to use constraints' regions to pinpoint the most blameable constraint, but I haven't found a straightforward way to get blame out of the constraint graph or region graph, given the cycles. constraints' location/span are helpful, but mostly once we're in proper diagnostic code and can look at types and the HIR and MIR and such. the compromise I've imagined is to pick a good-enough-but-maybe-not-best constraint to blame, and anything important enough to surface in addition to that can be picked out from the constraint path by diagnostics (though of course if there's some other way to get the blamed constraint as good as possible too, that'd be ideal).